Os 386 cancellation tokens

jared-improbable opened the pull request.

jared-improbable committed:
Added display name and player id to locator options
Fixed memory leak of futures
`UseInsecureConnection = true`

jared-improbable committed:
Implement members

jared-improbable committed:
Simplified internal locator connection api
Fixed protocol logging options not propagating in Locator flow

jared-improbable committed:
Better names
Properly plumb through UseInsecureConnection

jared-improbable committed:
Need the port too

jared-improbable committed:
Remove `AssociatePathWithClient` command

jared-improbable committed:
Default to receptionist-only connection flow
Remove `LogMessageOp` handling

jared-improbable committed:
Use worker type rather than worker attributes for write access

jared-improbable committed:
Removed `GetMetrics` command

jared-improbable committed:
Make `dev-token` an optional argument

jared-improbable committed:
Compile fix

jared-improbable committed:
Use built-in env var

jared-improbable committed:
Dbg

jared-improbable committed:
Fixed SQL to work with PSQL below 11.x

jared-improbable committed:
Add a simple Platform SDK wrapper
Managed `spatiald`, and simplifies starting and stopping local dpls
Added concrete (but incomplete) types for local dpl and spatialos.json files

jared-improbable committed:
Version bump

jared-improbable committed:
Format

jared-improbable committed:
Cleanup

jared-improbable committed:
CR Feedback

jared-improbable committed:
Code cleanup
Fixed disconnection not cancelling all outstanding command requests

jared-improbable committed:
Fixed being able to deadlock by `await`ing a command on a connection which has been disconnected

jared-improbable committed:
Proper async implmentations with progress and cancellation
Code cleanup

jared-improbable committed:
Error handling, more debug info

jared-improbable committed:
Process handle leak

jared-improbable committed:
Bit of debugging

jared-improbable committed:
Latest toolbelt

jared-improbable committed:
Workflows feedback: Try a sleep

jared-improbable committed:
Tracking ticket

jared-improbable committed:
Typo fixes

jared-improbable committed:
Version bumps

jared-improbable committed:
Version bump

jared-improbable committed:
Version

jared-improbable committed:
First pass on adding CancellationTokens

jared-improbable committed:
Dead code cleanup

jared-improbable committed:
Propagate task cancellation better
Allow connection attempts to be canceled
Switch back to `RunContinuationsAsynchronously` to avoid deadlocks
Version bump

jared-improbable committed:
Ergonomics improvement to avoid `null, null, null`
* Make all constructor parameters `default`
* Added special cases for `null` collection types
Don't generate `Update` objects and command responders for restricted components
Fixed exception propogation from the test helpers

jared-improbable committed:
Shuffled project info types into their own project
Added `WorkerConfig` type

jared-improbable committed:
Merged with master

jared-improbable committed:
Weird merge?

jared-improbable committed:
Weird merge

jared-improbable committed:
Project info generator

jared-improbable committed:
Comments

jared-improbable committed:
Fixed `Dispose` not being thread-safe

jamiebrynes7 commented on Improbable/Stdlib/Improbable.Stdlib.CSharpCodeGen/ProjectInfoGenerator.cs:
```diff
@@ -0,0 +1,59 @@
+        public string GenerateProjectInfo(string pathToSpatialOsJson, string ns = "SpatialOsProject")
+        {
+            var project = JsonConvert.DeserializeObject<ProjectConfig>(File.ReadAllText(pathToSpatialOsJson, Encoding.UTF8));
+            var configDir = Path.GetDirectoryName(pathToSpatialOsJson) ?? Environment.CurrentDirectory;
```
So `Path.GetDirectoryName` only returns null for if the directory is root (or is null). 

In these cases using `Environment.CurrentDirectory` feels wrong?

jamiebrynes7 commented on Improbable/Stdlib/Improbable.Stdlib.CSharpCodeGen/ProjectInfoGenerator.cs:
```diff
@@ -0,0 +1,59 @@
+
+            var layers = new HashSet<string>();
+            var workerTypes = new HashSet<string>();
+            var mapping = new Dictionary<string, List<string>>();
```
mapping of... what?

jared-improbable replied:
Oops, leftover from a previous experiment - this is unneeded now.

jamiebrynes7 commented on Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs:
```diff
@@ -43,41 +44,59 @@ public string WorkerId
+                        {
+                            cancellation.ThrowIfCancellationRequested();
+
+                            if (future.TryGet(out connection, 50))
```
Ugh this "future" should _really_ be a `Task<Connection>` and this polling should be pushed back into the SDK imo

jamiebrynes7 commented on Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs:
```diff
@@ -125,9 +148,18 @@ private static Task<WorkerConnection> ConnectAsync(ILocatorOptions options, Conn
                     using (var locator = new Locator(options.SpatialOsHost, options.SpatialOsPort, locatorParameters))
                     using (var future = locator.ConnectAsync(connectionParameters))
```
Some common code for polling the future that can be factored out. 


We might actually be able to write a generic extension method to transform it into a task.


```csharp
        public static Task<T> ToTask<T>(this Future<T> future, CancellationToken cancellation = default)
        {
            var tcs = new TaskCompletionSource<T>();
            cancellation.Register(() => tcs.TrySetCanceled(cancellation));

            Task.Run(() =>
            {
                while (true)
                {
                    cancellation.ThrowIfCancellationRequested();

                    if (future.TryGet(out var value, 50))
                    {
                        tcs.TrySetResult(value);
                        break;
                    }
                }
            }, cancellation);

            return tcs.Task;
        }
```

Or something like that

jared-improbable replied:
Nice, good idea

jamiebrynes7 commented on Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs:
```diff
@@ -535,15 +561,11 @@ public OpList GetOpList(TimeSpan timeout)
-
-                return EmptyOpList;
+                CancelCommands();
+                throw new Exception("Not connected to SpatialOS");
```
`InvalidOperationException`?

jamiebrynes7 commented on Improbable/Test/Improbable.Test/ProcessExtensions.cs:
```diff
@@ -1,125 +0,0 @@
-using System;
```
🎉

jared-improbable committed:
Proper thread safety for connection object
Removed the test project and just inlined the functionality into the local project

jared-improbable committed:
Cleanups

jared-improbable committed:
Refactored `Future<>` polling to extension method

jared-improbable committed:
Add a generic type for streaming events

jared-improbable committed:
More robust startup code

jared-improbable committed:
Try again

jamiebrynes7 commented on Improbable/Stdlib/Improbable.Stdlib/EventProcessor.cs:
```diff
@@ -0,0 +1,55 @@
+
+namespace Improbable.Stdlib
+{
+    public class EventProcessor<T> : IOpProcessor where T : struct
```
Is this the most specific generic constraint?

jared-improbable replied:
Yeah, there's no inheritance or anything else to go against

jamiebrynes7 commented on Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs:
```diff
@@ -88,41 +90,61 @@ public static async Task<Deployment> StartLocalAsync(StartDeploymentOptions star
 
-            progress?.Report($"Starting local deployment for '{project.ProjectName}' with snapshot '{startDeployment.SnapshotId}'...");
-            var response = client.CreateDeployment(new CreateDeploymentRequest
+            var retries = 10;
```
Nit: can be const
